pm_kext/col/verdict_cache.c

```diff
@@ -135,34 +156,55 @@ int teardownVerdictCache(VerdictCache *verdictCache) {
  * @return error code
  *
  */
-int addVerdict(VerdictCache *verdictCache, PortmasterPacketInfo *packetInfo, verdict_t verdict) {
-    if (!verdictCache || !packetInfo || !verdict) {
-        ERR("add_verdict NULL pointer exception verdictCache=0p%Xp, packetInfo=0p%Xp, verdict=0p%Xp ", verdictCache, packetInfo, verdict);
+int addVerdict(VerdictCache *verdictCache, PortmasterPacketInfo *packetInfo, verdict_t verdict, PortmasterPacketInfo **removedPacketInfo) {
+    if (verdictCache == NULL || packetInfo == NULL || verdict == 0) {
+        ERR("addVerdict NULL pointer exception verdictCache=0p%Xp, packetInfo=0p%Xp, verdict=0p%Xp ", verdictCache, packetInfo, verdict);
         return 1;
     }
 
-    VerdictCacheItem *newItem = _ALLOC(sizeof(VerdictCacheItem), 1);
-    if(!newItem) {
-        ERR("add_verdict tried to add NULL-Pointer verdict");
-        return 2;
-    }
+    counter++;
 
-    newItem->packetInfo = packetInfo;
-    newItem->verdict = verdict;
+    VerdictCacheItem *newItem = NULL;
 
-    // insert as first item
-    if (verdictCache->head != NULL) {
-        newItem->next = verdictCache->head;
-        verdictCache->head->prev = newItem;
+    VerdictCacheKey key = getCacheKey(packetInfo);
+    #pragma warning(suppress : 4127) //  warning C4127: conditional expression is constant -> if generated by macro
+    HASH_FIND(hh, verdictCache->map, &key, sizeof(VerdictCacheKey), newItem);
+    if(newItem != NULL) {
+        INFO("addVerdict packet was already in");
+        // already in
+        return 3;
     }
-    verdictCache->head = newItem;
 
-    // set tail if only item
-    if (verdictCache->tail == NULL) {
-        verdictCache->tail = newItem;
+    if(verdictCache->numberOfFreeItems > 0) {
+        newItem = &verdictCache->itemPool[verdictCache->maxSize - verdictCache->numberOfFreeItems];
+        verdictCache->numberOfFreeItems -= 1;
+    } else {
+        VerdictCacheItem *item = getOldestAccessTimeItem(verdictCache);
+        if(item == NULL) {
+            ERR("addVerdict failed to find free element");
+            return 2;
+        }
+        *removedPacketInfo = item->packetInfo;
+        resetItem(verdictCache, item);
+        newItem = item;
+    }
+
+    // Set key
+    newItem->key = key;
+    newItem->redirectKey = getCacheRedirectKey(packetInfo);
+    newItem->packetInfo = packetInfo;
+    newItem->verdict = verdict;
+    newItem->lastAccessed = counter;
+    HASH_ADD(hh, verdictCache->map, key, sizeof(VerdictCacheKey), newItem);
+
+    // insert only if we dont have already item with the same key
+    VerdictCacheItem *redirectItem = NULL;
+    #pragma warning(suppress : 4127) //  warning C4127: conditional expression is constant -> if generated by macro
+    HASH_FIND(hhRedirect, verdictCache->mapRedirect, &newItem->redirectKey, sizeof(VerdictCacheKey), redirectItem);
+    if(redirectItem == NULL) {
+        HASH_ADD(hhRedirect, verdictCache->mapRedirect, redirectKey, sizeof(VerdictCacheKey), newItem);
     }
```

Comment:
Would adding not simply override the entry? How do we update verdicts?

Also, is this done for _every_ packet, or just where we need redirection?